activesupport/lib/active_support/core_ext/class/attribute.rb

```diff
@@ -122,11 +123,37 @@ def class_attribute(*attrs, instance_accessor: true,
           methods << "silence_redefinition_of_method def #{name}?; !!self.#{name}; end"
         end
       end
+
+      if default.is_a?(Hash)
+        class_methods << inplace_updatable_class_attribute(name)
+      end
     end
 
     location = caller_locations(1, 1).first
     class_eval(["class << self", *class_methods, "end", *methods].join(";").tr("\n", ";"), location.path, location.lineno)
 
     attrs.each { |name| public_send("#{name}=", default) }
   end
+
+  private
+    def inplace_updatable_class_attribute(name)
+      update_method = name.to_s.sub(/\A(_*)/, "\\1inplace_update_")
+
+      <<~RUBY
+        def #{update_method}(key, value)
+          prev_value = #{name}[key]
+          #{name}[key] = value
+
+          descendants.each do |subclass|
+            next if subclass.#{name}.equal?(#{name})
+
+            if !subclass.#{name}.key?(key) || prev_value&.equal?(subclass.#{name}[key])
```

Comment:
This can produce different outcomes for flyweight / frozen objects versus other objects.  For example:

```ruby
A = Class.new
A.class_attribute :things, default: {}
B = Class.new(A)

A.inplace_update_things(:point, [0, 0])
B.things = B.things.dup
B.inplace_update_things(:point, [0, 0])
A.inplace_update_things(:point, [1, 1])
B.things[:point]  # => [0, 0]

A.inplace_update_things(:x, 0)
B.things = B.things.dup
B.inplace_update_things(:x, 0)
A.inplace_update_things(:x, 1)
B.things[:x]  # => 1
```

My approach to fix this was to [track overridden keys](https://github.com/rails/rails/pull/39487/commits/f9483f36c8e4746b55d944c33fb3be3acd7b9f46#diff-531e71a9d5fdc8181f38ce4416553694R136-R141).

Reply:
I know this does not perfectly detect overrides.
But I suppose this works well enough as a small utility that solves such like internal problems.

Reply:
Note that this small utility is focused to avoid allocation as much as possible.

Reply:
Yes, I am not a fan of extra allocation either.  But I felt it was necessary because even if such problems are unlikely, they would be hard to diagnose when they did occur.  Also, because it is implemented as an automatic feature of `class_attribute`, it seems more public, and so I wanted to make it more foolproof.

Reply:
I added https://github.com/rails/rails/pull/39487/commits/ff3e8f4f92a56c3b8925d9cbb373eb5658d428ad which should prevent Set allocation when the superclass's `class_attribute` value is `empty?` or undefined.  If the update method is called on the superclass later (and its `class_attribute` value becomes non-empty), the subclass's Set will be allocated and properly back-filled.